### PR TITLE
Add Adafactor foreach impl

### PR DESCRIPTION
**Pull request by janeyx99**
This PR adds the foreach impl for Adafactor knowing that there are many ways to improve its runtime perf today (by adding more foreach support). After this PR:
- we have a foreach flag for Adafactor
- It is NOT the default. Why not? It is only slightly faster + uses O(n) more memory where n is the number of params in your max param group. People tend to use Adafactor for memory efficiency.

Next steps:
- make torch.compile possible on it
- make it faster (by adding more foreach apis)

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #132336
* #133360

### CODE REVIEW

File: torch/optim/_adafactor.py [hunk 1]
```diff
@@ -395,6 +407,186 @@ def _single_tensor_adafactor(
...
+            ), "row_var and col_var should be defined when grad is multidimensional"
+            # same as (g * g).mean(dim=-1) w/o materializing an intermediate size g
+            row_means = [
+                torch.norm(grad, dim=-1, keepdim=True) for grad in device_grads
+            ]
+            torch._foreach_mul_(row_means, row_means)
+            torch._foreach_div_(row_means, [grad.size(-1) for grad in device_grads])  # type: ignore[union-attr]
+            torch._foreach_mul_(device_row_vars, beta2_ts)  # type: ignore[arg-type]
+            torch._foreach_mul_(row_means, one_minus_beta2_ts)
+            torch._foreach_add_(device_row_vars, row_means)  # type: ignore[arg-type]
```
Comment (janeyx99) [inline torch/optim/_adafactor.py:534]: in the future would be a 
```
torch._foreach_lerp_(device_row_vars, row_means, one_minus_beta2_ts)
```

if we had ScalarList support for _foreach_lerp 3rd arg

File: torch/optim/_adafactor.py [hunk 2]
```diff
@@ -395,6 +407,186 @@ def _single_tensor_adafactor(
...
+
+            # same as (g * g).mean(dim=-2) w/o materializing an intermediate size g
+            col_means = [
+                torch.norm(grad, dim=-2, keepdim=True) for grad in device_grads
+            ]
+            torch._foreach_mul_(col_means, col_means)
+            torch._foreach_div_(col_means, [grad.size(-2) for grad in device_grads])  # type: ignore[union-attr]
+            torch._foreach_mul_(device_col_vars, beta2_ts)  # type: ignore[arg-type]
+            torch._foreach_mul_(col_means, one_minus_beta2_ts)
+            torch._foreach_add_(device_col_vars, col_means)  # type: ignore[arg-type]
```
Comment (janeyx99) [inline torch/optim/_adafactor.py:545]: `torch._foreach_lerp_(device_col_vars, col_means, one_minus_beta2_ts)`

File: torch/optim/_adafactor.py [hunk 3]
```diff
@@ -395,6 +407,186 @@ def _single_tensor_adafactor(
...
+            del row_var_means
+        else:
+            assert (
+                device_variances[0] is not None
+            ), "variance should be defined when grad is a vector"
+
+            grads_squared = torch._foreach_mul(device_grads, device_grads)  # type: ignore[arg-type]
+            torch._foreach_mul_(device_variances, beta2_ts)  # type: ignore[arg-type]
+            torch._foreach_mul_(grads_squared, one_minus_beta2_ts)
+            torch._foreach_add_(device_variances, grads_squared)  # type: ignore[arg-type]
```
Comment (janeyx99) [inline torch/optim/_adafactor.py:566]: `torch._foreach_lerp_(device_variances, grads_squared, one_minus_beta2_ts)`

File: torch/optim/_adafactor.py [hunk 4]
```diff
@@ -395,6 +407,183 @@ def _single_tensor_adafactor(
...
+        if weight_decay != 0:
+            torch._foreach_mul_(device_params, 1 - lr * weight_decay)  # type: ignore[arg-type]
+
+        if is_multidim:
+            assert (
+                device_row_vars[0] is not None and device_col_vars[0] is not None
+            ), "row_var and col_var should be defined when grad is multidimensional"
+            # same as (g * g).mean(dim=-1) w/o materializing an intermediate size g
+            row_means = [
+                torch.norm(grad, dim=-1, keepdim=True) for grad in device_grads
```
Comment (janeyx99) [inline torch/optim/_adafactor.py:528]: no foreach norm support for this type of norm

File: torch/optim/_adafactor.py [hunk 5]
```diff
@@ -395,6 +407,183 @@ def _single_tensor_adafactor(
...
+            torch._foreach_mul_(col_means, one_minus_beta2_ts)
+            torch._foreach_add_(device_col_vars, col_means)  # type: ignore[arg-type]
+            del col_means
+
+            var_estimates = [
+                row_var @ col_var  # type: ignore[operator]
+                for row_var, col_var in zip(device_row_vars, device_col_vars)
+            ]
+            row_var_means = [
+                row_var.mean(dim=-2, keepdim=True) for row_var in device_row_vars  # type: ignore[union-attr]
```
Comment (janeyx99) [inline torch/optim/_adafactor.py:553]: no foreach mean

File: torch/optim/_adafactor.py [hunk 6]
```diff
@@ -395,6 +407,183 @@ def _single_tensor_adafactor(
...
+            ]
+            torch._foreach_mul_(col_means, col_means)
+            torch._foreach_div_(col_means, [grad.size(-2) for grad in device_grads])  # type: ignore[union-attr]
+            torch._foreach_mul_(device_col_vars, beta2_ts)  # type: ignore[arg-type]
+            torch._foreach_mul_(col_means, one_minus_beta2_ts)
+            torch._foreach_add_(device_col_vars, col_means)  # type: ignore[arg-type]
+            del col_means
+
+            var_estimates = [
+                row_var @ col_var  # type: ignore[operator]
```
Comment (janeyx99) [inline torch/optim/_adafactor.py:549]: no foreach mm lol, probably the bulk of the work

File: torch/optim/_adafactor.py [hunk 7]
```diff
@@ -395,6 +407,183 @@ def _single_tensor_adafactor(
...
+            torch._foreach_add_(device_variances, grads_squared)  # type: ignore[arg-type]
+            del grads_squared
+
+            # avoid writing into variance during update
+            var_estimates = [v.clone() for v in device_variances]  # type: ignore[union-attr]
+
+        # square the eps1 as we sqrt after to keep eps1's magnitude
+        torch._foreach_clamp_min_(var_estimates, eps1 * eps1)
+        torch._foreach_sqrt_(var_estimates)
+        torch._foreach_reciprocal_(var_estimates)
```
Comment (janeyx99) [inline torch/optim/_adafactor.py:575]: would benefit from a foreach_rsqrt

File: torch/optim/_adafactor.py [hunk 8]
```diff
@@ -395,6 +407,183 @@ def _single_tensor_adafactor(
...
+        torch._foreach_mul_(var_estimates, device_grads)  # type: ignore[arg-type]
+        updates = var_estimates
+
+        alphas = [
+            -a / (max(1.0, update.norm(2).item() / ((update.numel() ** 0.5) * d)))
+            for a, update in zip(alphas, updates)
+        ]
+        torch._foreach_mul_(updates, alphas)
+        torch._foreach_add_(device_params, updates)  # type: ignore[arg-type]
```
Comment (janeyx99) [inline torch/optim/_adafactor.py:584]: would be nice to have a foreach_add where the alphas could be a scalarlist

File: torch/optim/_adafactor.py [hunk 9]
```diff
@@ -395,6 +407,183 @@ def _single_tensor_adafactor(
...
+        # Update steps
+        # If steps are on CPU, foreach will fall back to the slow path, which is a for-loop calling t.add(1) over
+        # and over. 1 will then be wrapped into a Tensor over and over again, which is slower than if we just
+        # wrapped it once now. The alpha is required to assure we go to the right overload.
+        if not torch._utils.is_compiling() and device_state_steps[0].is_cpu:
+            torch._foreach_add_(
+                device_state_steps, torch.tensor(1.0, device="cpu"), alpha=1.0  # type: ignore[arg-type]
+            )
+        else:
+            torch._foreach_add_(device_state_steps, 1)  # type: ignore[arg-type]
```
Comment (albanD) [inline torch/optim/_adafactor.py:503]:
```suggestion
            torch._foreach_add_(device_state_steps, 1.)  # type: ignore[arg-type]
```

?

File: torch/optim/_adafactor.py (hunk 1 above)
Comment (albanD) [reply torch/optim/_adafactor.py:534]: Do you expect to do these before merging the PR?
Should these improvements be recapped in an issue?

**Review by albanD — COMMENTED** [review]
Sounds pretty good!
Curious what's the plan for all the future improvements

**Comment by janeyx99** [comment]
@albanD I'm planning to encapsulate all the action items in an issue before landing this PR. Including perf wins, compile support, etc.

**Comment by janeyx99** [comment]
Perf tracker with all issues: https://github.com/pytorch/pytorch/issues/133367

**Review by albanD — APPROVED** [review]
Sounds good!

**Pull request closed without merging**